index.bs

```diff
@@ -2857,14 +2857,48 @@ Note: The {{CollectedClientData}} may be extended in the future. Therefore it's
 
     The {{CollectedClientData}} structure is used by the client to compute the following quantities:
 
-    : <dfn dfn>JSON-serialized client data</dfn>
-    :: This is the result of [=serialize JSON to bytes|JSON-serializing to bytes=] a
-        {{CollectedClientData}} dictionary.
+    : <dfn dfn>JSON-compatible serialization of client data</dfn>
+    :: This is the result of performing the [JSON-compatible serialization algorithm](#clientdatajson-serialization) on the {{CollectedClientData}} dictionary.
 
     : <dfn dfn>Hash of the serialized client data</dfn>
-    :: This is the hash (computed using SHA-256) of the [=JSON-serialized client data=], as constructed by the client.
+    :: This is the hash (computed using SHA-256) of the [=JSON-compatible serialization of client data=], as constructed by the client.
 </div>
 
+#### Serialization #### {#clientdatajson-serialization}
+
+The serialization of the {{CollectedClientData}} is a subset of the algorithm for [=serialize JSON to bytes|JSON-serializing to bytes=]. I.e. it produces a valid JSON encoding of the {{CollectedClientData}} but also provides additional structure that may be exploited by verifiers to avoid integrating a full JSON parser.
+
+The serialization algorithm works by appending successive byte strings to an, initially empty, partial result until the complete result is obtained. If a step in the following algorithm suggests appending a string, then the byte string to append is formed by [UTF-8 encoding](https://encoding.spec.whatwg.org/#utf-8-encode) that string.
+
+1. Append &lsquo;`{"type":`&rsquo;.
+1. Append [=CCDToString=]({{CollectedClientData/type}}).
+1. Append &lsquo;`,"challenge":`&rsquo;.
+1. Append [=CCDToString=]({{CollectedClientData/challenge}}).
+1. Append &lsquo;`,"origin":`&rsquo;.
+1. Append [=CCDToString=]({{CollectedClientData/origin}}).
+1. Append &lsquo;`,"crossOrigin":`&rsquo;.
+1. If {{CollectedClientData/crossOrigin}} is not included, or is `false`:
+    1. Append &lsquo;`false`&rsquo;.
+1. Otherwise:
+    1. Append &lsquo;`true`&rsquo;.
+1. Create a temporary copy of the {{CollectedClientData}} and remove the fields `type`, `challenge`, `origin`, and `crossOrigin` (if present). Invoke [=serialize JSON to bytes=] on it to produce a byte string `remainder`.
+1. If `remainder` is equal to 0x7b7d (i.e. "{}") then:
+    1. Append &lsquo;`}`&rsquo;.
+1. Otherwise:
+    1. Append &lsquo;`,`&rsquo;.
+    1. Remove the leading byte from `remainder`.
+    1. Append `remainder` to the partial result.
+
+The function <dfn>CCDToString</dfn> is used in the above algorithm and is defined as:
```

Comment:
this function seems to have an implied accumulator variable that perhaps oughta be explicit.

Reply:
Have made accumulator values explicit.